src/nrnpython/inithoc.cpp

```diff
@@ -49,21 +50,169 @@ extern int nrn_nobanner_;
 extern int ivocmain(int, char**, char**);
 extern int nrn_main_launch;
 
-#ifdef NRNMPI
-
-static const char* argv_mpi[] = {"NEURON", "-mpi", "-dll", 0};
-static int argc_mpi = 2;
-
-#endif
-
-static const char* argv_nompi[] = {"NEURON", "-dll", 0};
-static int argc_nompi = 1;
-
 #if USE_PTHREAD
 #include <pthread.h>
 static pthread_t main_thread_;
 #endif
 
+/**
+ * Manage argc,argv for calling ivocmain
+ * add_arg(...) will only add if name is not already in the arg list
+ * returns 1 if added, 0 if not
+*/
+static size_t arg_size;
+static int argc;
+static char** argv;
+static int add_arg(const char* name, const char* value) {
+  if (size_t(argc + 2) >= arg_size) {
+    arg_size += 10;
+    argv = (char**)realloc(argv, arg_size*sizeof(char*));
+  }
+  // Don't add if already in argv
+  for (int i=1; i < argc; ++i) {
+    if (strcmp(name, argv[i]) == 0) {
+      return 0;
+    }
+  }
+  argv[argc++] = strdup(name);
+  if (value) {
+    argv[argc++] = strdup(value);
+  }
+  return 1;
+}
+
+/**
+ * Return 1 if string, 0 otherwise.
+*/
+static int is_string(PyObject* po) {
+  if (PyUnicode_Check(po) || PyBytes_Check(po)) {
+    return 1;
+  }
+  return 0;
+}
+
+/**
+ * Add all name:value from __main__.neuron_options dict if exists
+ * to the argc,argv for calling ivocmain
+ * Note: if value is None then only name is added to argv.
+ * The special "-print-options" name is not added to argv but
+ * causes a return value of 1. Otherwise the return value is 0.
+*/
+static int add_neuron_options() {
+  PyObject* modules = PyImport_GetModuleDict();
+  PyObject* module = PyDict_GetItemString(modules, "__main__");
+  int rval = 0;
+  if (!module) {
+    PyErr_Clear();
+    PySys_WriteStdout("No __main__  module\n");
+    return rval;
+  }
+  PyObject* neuron_options = PyObject_GetAttrString(module, "neuron_options");
+  if (!neuron_options) {
+    PyErr_Clear();
+    return rval;
+  }
+  if (!PyDict_Check(neuron_options)) {
+    PySys_WriteStdout("__main__.neuron_options is not a dict\n");
+    return rval;
+  }
+  PyObject *key, *value;
+  Py_ssize_t pos = 0;
+  while(PyDict_Next(neuron_options, &pos, &key, &value)) {
+    if (!is_string(key) || (!is_string(value) && value != Py_None)) {
+     PySys_WriteStdout("A neuron_options key:value is not a string:string or string:None\n");
+     continue;
+    }
+    Py2NRNString skey(key);
+    Py2NRNString sval(value);
+    if (strcmp(skey.c_str(), "-print-options") == 0) {
+      rval = 1;
+      continue;
+    }
+    add_arg(skey.c_str(), sval.c_str());
+  }
+  return rval;
+}
+
+/**
+ * Space separated options. Must handle escaped space, '...' and "...".
+ * Return 1 if contains a -print-options (not added to options)
+*/
+
+static int add_space_separated_options(const char* str) {
```

Comment:
I have one suggestion. Not familiar with all options and values, but was thinking it would be simpler if we had another delimiter, say `$` (or even `$$` if the former impacts in any way) and used `strtok` to pull out options. We could then use `=` if the option has a value, and look for a first occurrence that will tell us where to cut -> no need to escape. 

Something along these lines: 
```
#include <string.h>
#include <stdio.h>

int main () {
   char str[] = "-nogui$$-background=#ffffffff$$-stroption=this=that;that=this";
   const char s[3] = "$$";
   const char e[2] = "=";
   char *token;
   
   /* get the first token */
   token = strtok(str, s);
   
   /* walk through other tokens */
   while( token != NULL ) {
      printf( "%s\n", token );
      char * optionValues; 
      /* look for first occurrence of '=' */
      if ( (optionValues = strstr(token, e)) != NULL) {
          printf("\t%s\n", optionValues +1);
      }
    
      token = strtok(NULL, s);
   }
   
   return(0);
}
```

```
-nogui
-background=#ffffffff
	#ffffffff
-stroption=this=that;that=this
	this=that;that=this
```

Reply:
@alexsavulescu I hate to get too far away from normal option syntax. On all machines, space separated arguments are what a person types on the command line to launch a program.  And if args  contain spaces or other special shell characters, e.g. ';' or '#', then they need to be escaped with `\` or the entire arg enclosed in "..." or '....' depending on how the entire environment string is enclosed in the export statement. And if one gets away from space separated args then one cannot test with
```nrniv $NEURON_MODULE_OPTIONS```

Reply:
I see. Didn't think about `nrniv $NEURON_MODULE_OPTIONS`